src/pip/_internal/network/download.py

```diff
@@ -250,6 +253,77 @@ def _attempt_resumes_or_redownloads(
             os.remove(download.output_file.name)
             raise IncompleteDownloadError(download)
 
+        # If we successfully completed the download via resume, manually cache it
+        # as a complete response to enable future caching
+        if download.reattempts > 0:
+            self._cache_resumed_download(download, first_resp)
+
+    def _cache_resumed_download(
+        self, download: _FileDownload, original_response: Response
+    ) -> None:
+        """
+        Manually cache a file that was successfully downloaded via resume retries.
+
+        cachecontrol doesn't cache 206 (Partial Content) responses, since they
+        are not complete files. This method manually adds the final file to the
+        cache as though it was downloaded in a single request, so that future
+        requests can use the cache.
+        """
+        url = download.link.url_without_fragment
+        if url.startswith("https://"):
+            adapter = self._session.adapters["https://"]
+        elif url.startswith("http://"):
+            adapter = self._session.adapters["http://"]
+        else:
+            return
+
+        # Check if the adapter is the CacheControlAdapter (i.e. caching is enabled)
+        if not isinstance(adapter, CacheControlAdapter):
+            logger.debug(
+                "Skipping resume download caching: no cache controller for %s", url
+            )
+            return
+
+        # Check SafeFileCache is being used
+        if not isinstance(adapter.cache, SafeFileCache):
+            logger.debug(
+                "Skipping resume download caching: "
+                "cache doesn't support separate body storage"
+            )
+            return
```

Comment:
```suggestion
        assert isinstance(adapter.cache, SafeFileCache), "separate body cache not in use!"
```

I'm pretty sure all instances of `CacheControlAdapter` are guaranteed to be using the `SafeFileCache`. I'd rather that we crash loudly if this isn't true so we can fix it.

Reply:
Applied